portable/IAR/ARM_CM23/non_secure/port.c

```diff
@@ -1046,19 +1056,34 @@ void vPortEndScheduler( void ) /* PRIVILEGED_FUNCTION */
 		 * the stack region has already been configured. */
 		if( ulStackDepth > 0 )
 		{
-			/* Define the region that allows access to the stack. */
-			ulRegionStartAddress = ( ( uint32_t ) pxBottomOfStack ) & portMPU_RBAR_ADDRESS_MASK;
+			ulRegionStartAddress = ( uint32_t ) pxBottomOfStack;
 			ulRegionEndAddress = ( uint32_t ) pxBottomOfStack + ( ulStackDepth * ( uint32_t ) sizeof( StackType_t ) ) - 1;
-			ulRegionEndAddress  &= portMPU_RLAR_ADDRESS_MASK;
 
-			xMPUSettings->xRegionsSettings[ 0 ].ulRBAR =	( ulRegionStartAddress ) |
-															( portMPU_REGION_NON_SHAREABLE ) |
-															( portMPU_REGION_READ_WRITE ) |
-															( portMPU_REGION_EXECUTE_NEVER );
+			/* If the stack is within the privileged SRAM, do not protect it
+			 * using a separate MPU region. This is needed because privileged
+			 * SRAM is already protected using an MPU region and ARMv8-M does
+			 * not allow overlapping MPU regions. */
+			if( ulRegionStartAddress >= ( uint32_t ) __privileged_sram_start__ &&
+				ulRegionEndAddress <= ( uint32_t ) __privileged_sram_end__ )
+			{
+				xMPUSettings->xRegionsSettings[ 0 ].ulRBAR = 0;
+				xMPUSettings->xRegionsSettings[ 0 ].ulRLAR = 0;
+			}
+			else
+			{
+				/* Define the region that allows access to the stack. */
+				ulRegionStartAddress &= portMPU_RBAR_ADDRESS_MASK;
+				ulRegionEndAddress  &= portMPU_RLAR_ADDRESS_MASK;
+
+				xMPUSettings->xRegionsSettings[ 0 ].ulRBAR =	( ulRegionStartAddress ) |
+																( portMPU_REGION_NON_SHAREABLE ) |
+																( portMPU_REGION_READ_WRITE ) |
+																( portMPU_REGION_EXECUTE_NEVER );
 
-			xMPUSettings->xRegionsSettings[ 0 ].ulRLAR =	( ulRegionEndAddress ) |
-															( portMPU_RLAR_ATTR_INDEX0 ) |
-															( portMPU_RLAR_REGION_ENABLE );
+				xMPUSettings->xRegionsSettings[ 0 ].ulRLAR =	( ulRegionEndAddress ) |
+																( portMPU_RLAR_ATTR_INDEX0 ) |
+																( portMPU_RLAR_REGION_ENABLE );
+			}

```

Comment:
How is a stack overflow detected on the M23?  The M33 has the stack limit registers.

Reply:
A separate MPU region is used to grant a task access to its stack. If a stack overflows its stack, Memory Fault will be generated. 

There is a corner case though - if the stack is placed right next to a memory region to which the task has Read-Write access, stack overflow wont trigger a memory fault as the task will still write to the memory to which it has write permission. 

```
   +-------------------------------+---------------------------------+
   |                               |                                 |
   |     Region with RW access     |            Task Stack           |
   |                               |                                 |
   +-------------------------------+---------------------------------+
```

The above mentioned corner case only exists in M23 as stack limit registers will trigger fault in M33.